BankManagerConsoleApp/BankManagerConsoleApp/Bank.swift

```diff
@@ -38,34 +39,43 @@ struct Bank {
                 break
             }
             
-            work(for: customer)
+            workManager(for: customer)
         }
         
+        workingGroup.wait()
+        
         let finishDate = Date().timeIntervalSince(startDate)
         
         print("업무가 마감되었습니다. 오늘 업무를 처리한 고객은 총 \(numberOfCustomer)명이며, 총 업무시간은 \(finishDate.applyNumberFormatter())초입니다.")
         run()
     }
     
-    mutating private func receiveNumberOfCustomers() -> Int {
+    private func receiveNumberOfCustomers() -> Int {
         let numberOfCustomer = Int.random(in: BankOption.rangeOfCustomer)
         
         for count in 1...numberOfCustomer {
-            customers.enqueue(Customer(waitingNumber: count))
+            guard let randomBanking = Banking.allCases.randomElement() else { return count - 1 }
+            
+            customers.enqueue(Customer(waitingNumber: count, banking: randomBanking))
         }
         
         return numberOfCustomer
     }
     
-    mutating private func work(for customer: Customer) {
-        managers[0].customer = customer
-        
-        managers[0].work()
-    }
-    
-    mutating private func addManager() {
-        for _ in 1...BankOption.numberOfManager {
-            self.managers.append(BankManager())
+    private func workManager(for customer: Customer) {
+        switch customer.banking {
+        case .deposit:
+            DispatchQueue.global().async(group: workingGroup) { [weak self] in
+                self?.depositManager.wait()
```

Comment:
글로벌큐를 사용하신 이유는 무엇일까요 ? 
데드락은 무엇일까요 ?

Reply:
### 글로벌 큐를 사용한 이유
[DispatchQueue](https://developer.apple.com/documentation/dispatch/dispatchqueue)
> Another way that apps consume too many threads is by creating too many private concurrent dispatch queues. Because each dispatch queue consumes thread resources, creating additional concurrent dispatch queues exacerbates the thread consumption problem. Instead of creating private concurrent queues, submit tasks to one of the global concurrent dispatch queues.

- 위의 공식문서에서 제시한대로, 커스텀한 동시 디스패치 큐를 너무 많이 생성하면 스레드 리소스를 많이 소비하게 된다고 합니다.
- 커스텀 동시 큐를 생성하는 대신 global 큐에서 작업을 하는 것이 좋다고하여 글로벌 큐를 사용하였습니다.

### Deadlock
- `Deadlock`(교착상태) : 두 개 이상의 프로세스에서 서로 끝나기만을 기다려 프로세스가 멈춰버리는 문제
- Serial Queue에서 `sync`의 특성때문에 발생할 수 있는 문제입니다.
- 아래의 예제에서는 queue 프로퍼티가 queue.sync 작업으로 인해 block-wait 상태가 되고, queue.sync는 queue.async 내에서 실행되고 있던 코드이기 때문에 queue.async의 Block-wait이 끝나기를 기다려 교착상태가 발생합니다.

``` swift
let queue = DispatchQueue(label: "label")
queue.async {
    queue.sync {
        // queue가 잠겨있기 때문에 이 부분은 실행할 수 없게된다.
        for _ in 1...5 {
            print("😀😀😀😀😀")
            sleep(1)
        }
    }
}
```

- `main.sync`만 사용할 때도 교착 상태에 빠집니다. main 스레드는 sync가 끝나기를, sync는 main 스레드의 Block-wait이 끝나기를 기다려 프로세스가 멈춰버립니다.